### PR TITLE
Change pause logic to use bitmask

### PR DESCRIPTION
**Pull request by Chuongv**
Let me know if I missed anything.

### CODE REVIEW

**Comment by dvor** [comment]
Have you finished with this pull request or it is in progress?

_(if you like you can edit title to be `WIP:  Change pause logic to use bitmask` for not-ready pull requests)_

**Comment by Chuongv** [comment]
I thought I updated it. Is it still showing for you "WIP"?

**Comment by dvor** [comment]
No, it isn't "WIP". Will review it then. :-)

**Comment by Chuongv** [comment]
Yea it should be ready now. I had missed a few edge cases when I was testing earlier. But should be good now.

File: Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m [hunk 1]
```diff
@@ -246,11 +236,45 @@ - (void)updateCall:(OCTCall *)call withStatus:(OCTCallStatus)status
...
+        [self.audioEngine stopAudioFlow:nil];
+    }
+    else if (! pause &&  (previousPausedStatus == OCTCallPausedStatusByUser)) {
+        [self.audioEngine startAudioFlow:nil];
+        [self.timer startTimerForCall:call];
+        self.audioEngine.friendNumber = friend.friendNumber;
+    }
+    else if (! pause) {
+        self.audioEngine.friendNumber = friend.friendNumber;
+    }
```
Comment (dvor) [inline Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m:277]: Tried to refactor this method a bit to reduce number of `else if` and used variables in the code.

```
- (void)putOnPause:(BOOL)pause call:(OCTCall *)call
{
    OCTRealmManager *realmManager = [self.dataSource managerGetRealmManager];
    BOOL wasPaused = call.pausedStatus != OCTCallPausedStatusNone;

    if (pause) {
        if (! wasPaused) {
            [self.timer stopTimer];
            [self.audioEngine stopAudioFlow:nil];
        }
    }
    else {
        OCTFriend *friend = [call.chat.friends firstObject];
        self.audioEngine.friendNumber = friend.friendNumber;

        if (call.pausedStatus & OCTCallPausedStatusByUser) {
            [self.audioEngine startAudioFlow:nil];
            [self.timer startTimerForCall:call];
        }
    }

    [realmManager updateObject:call withBlock:^(OCTCall *callToUpdate) {
        if (pause) {
            callToUpdate.pausedStatus |= OCTCallPausedStatusByUser;
            callToUpdate.onHoldStartInterval = callToUpdate.onHoldStartInterval ?: [[NSDate date] timeIntervalSince1970];
        }
        else {
            callToUpdate.pausedStatus &= ~OCTCallPausedStatusByUser;
            callToUpdate.onHoldStartInterval = nil;
        }
    }];
}
```

Comment (dvor) [inline Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m:277]: Also replaced `previousPausedStatus == OCTCallPausedStatusByUser` with `previousPausedStatus & OCTCallPausedStatusByUser`. I think it should be there, right?

File: Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m [hunk 2]
```diff
@@ -349,7 +380,26 @@ - (void)toxAV:(OCTToxAV *)toxAV callStateChanged:(OCTToxAVCallState)state friend
...
 
-    [self updateCall:call withState:state];
+    OCTCallPausedStatus pauseStatus = call.pausedStatus;
+
+    if ((pauseStatus == OCTCallPausedStatusNone) && (state == OCTToxAVFriendCallStatePaused)) {
+        [self.audioEngine stopAudioFlow:nil];
+        [self.timer stopTimer];
+    }
+
+    if ((pauseStatus == OCTCallPausedStatusByFriend) && (state != OCTToxAVFriendCallStatePaused)) {
```
Comment (dvor) [inline Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m:390]: I think `pauseStatus & OCTCallPausedStatusByFriend` should go here

**Comment by dvor** [comment]
Hm, `OCTSubmanagerCalls` is getting complicated. Different states are changed in lots of places. You don't mind if I'll try to refactor it a little bit? It would be great if you could provide more unit tests so I'll be sure I won't break anything.

**Comment by Chuongv** [comment]
> Hm, OCTSubmanagerCalls is getting complicated. Different states are changed in lots of places. You don't mind if I'll try to refactor it a little bit? It would be great if you could provide more unit tests so I'll be sure I won't break anything.

Feel free to do so :smile: I don't have any unit tests done yet, but you can use https://github.com/Antidote-for-Tox/Antidote/tree/chuongv/improveSwitchingBetweenCalls in the meantime to test. Just have to modify the podfile to point to the correct remote branch of objcTox.

I'll work on the unit tests in the meantime.

File: Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m (hunk 1 above)
Comment (Chuongv) [inline Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m:277]: > Also replaced previousPausedStatus == OCTCallPausedStatusByUser with previousPausedStatus & OCTCallPausedStatusByUser. I think it should be there, right?

Hmm, no you can't do that there. You want to check that _only_ `OCTCallPausedStatusByUser` is part of the bitmask, so you know whether or not to start the audioEngine back up again.

Comment (dvor) [inline Classes/Private/Manager/Submanagers/OCTSubmanagerCalls.m:277]: Oh, yeah, I see now. Missed that.

**Comment by Chuongv** [comment]
Closing this one in favor of https://github.com/Antidote-for-Tox/objcTox/pull/68

I won't be making any changes if you are going to refactor. If not, let me know and I can look into refactoring the logic, since it is a little messy.

**Pull request closed without merging**